### PR TITLE
[RLlib] DDPG re-factor to fit into RLlib's functional algorithm builder API.

### PR DESCRIPTION
**Pull request by sven1977**
DDPG is the only remaining "major" algorithm in RLlib that does not use the functional algo builder API (through the factory functions: build_tf_policy, build_torch_policy). This PR translates DDPG to use this API.
NOTE: This will make it much faster to eventually translate DDPG to support PyTorch as well.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by ericl** [comment]
How's MuJoCo perf looking? We should probably try out at least the existing tuned configs for TD3 and DDPG.

**Comment by sven1977** [comment]
Sure ...

**Comment by sven1977** [comment]
Learning performance is stable: Ran HalfCheetah on 1 GPU with DDPG:
```
== Status ==
Memory usage on this node: 5.7/240.1 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/32 CPUs, 0/2 GPUs, 0.0/157.96 GiB heap, 0.0/49.46 GiB objects
Result logdir: /home/ubuntu/ray_results/halfcheetah-ddpg
Number of trials: 2 (2 TERMINATED)
+---------------------------+------------+-------+---------------+--------+------------------+--------+----------+
| Trial name                | status     | loc   | use_pytorch   |   iter |   total time (s) |     ts |   reward |
|---------------------------+------------+-------+---------------+--------+------------------+--------+----------|
| DDPG_HalfCheetah-v2_00000 | TERMINATED |       | False         |    178 |          2447.96 | 178000 | 2013.23  |
```

File: rllib/agents/ddpg/ddpg_policy.py [hunk 1]
```diff
@@ -25,519 +31,389 @@
...
+    #    self.sess,
+    #    #obs_input=self.cur_observations,
+    #    sampled_action=self.output_actions,
+    #    loss=self.actor_loss + self.critic_loss,
+    #    loss_inputs=self.loss_inputs,
+    #    update_ops=q_batchnorm_update_ops + policy_batchnorm_update_ops,
+    #    explore=explore,
+    #    dist_inputs=self._distribution_inputs,
+    #    dist_class=Deterministic,
+    #    timestep=timestep)
```
Comment (ericl) [inline rllib/agents/ddpg/ddpg_policy.py:276]: remove this?

**Review by ericl — APPROVED** [review]

**Comment by sven1977** [comment]
Hey @ericl Could you merge? Tests are all ok, except for the known broken temp-file problem in `rllib/tests/test_io.py`. Thanks.

**Pull request merged**